test/src/test/java/org/corfudb/runtime/checkpoint/CheckpointSmokeTest.java

```diff
@@ -456,6 +462,293 @@ public void checkpointWriterInterleavedTest() throws Exception {
+                .getSnapshotTimestamp();
+        try {
+            cpw.startCheckpoint(snapshot);
+            cpw.appendObjectState(m.entryStream());
+            cpw.finishCheckpoint();
+
+            // Instantiate new runtime & map.
+            setRuntime();
+            Map<String, String> m2 = instantiateStringMap(streamName);
+            for (int i = 0; i < numKeys; i++) {
+                assertThat(m2.get(keyPrefix + i)).describedAs("get " + i)
+                        .isEqualTo(m.get(keyPrefix + i) + mutationSuffix);
+            }
+        } finally {
+            r.getObjectsView().TXEnd();
+        }
+    }
+
+    /**
+     * CheckpointWriter has BATCH_THRESHOLD_PERCENTAGE which decreases the
+     * writeSizeLimit to BATCH_THRESHOLD_PERCENTAGE * writeSizeLimit.
+     * This test checks that a new checkpointEntry is created when the
+     * total size of SMREntries exceeds the decreased writeSizeLimit.
+     */
+    @Test
+    @SuppressWarnings("checkstyle:magicnumber")
+    public void checkpointWriterSizeLimitMarginTest() throws Exception {
+        final String streamName = "mystream6";
+        final UUID streamId = CorfuRuntime.getStreamID(streamName);
+        final String keyPrefix = "a-prefix";
+        final String author = "Me, myself, and I";
+        // max write size set to 25 MB
+        final int maxWriteSize = 25 << 20;
+
+        StreamingMap<String, String> m = instantiateStringMap(streamName);
+
+        // set the max write size to 25 MB
+        getRuntime().getParameters().setMaxWriteSize(maxWriteSize);
+        getRuntime().getParameters().setCodecType(Codec.Type.NONE);
+        CheckpointWriter cpw = new CheckpointWriter(getRuntime(), streamId, author, (CorfuTable) m);
+        cpw.setSerializer(serializer);
+        final double batchThresholdPercentage = cpw.getBATCH_THRESHOLD_PERCENTAGE();
+
+        int i = 0;
+        int numBytesPerCheckpointEntry = 0;
+        // each entry is 1 MB
+        String payload = getRandomStringOfSize(1 << 20);
+        int nextSmrEntrySize = getSerializedSMREntrySize(keyPrefix + i, payload,
+                cpw.getKeyMutator(), cpw.getValueMutator(), cpw.getSerializer());
+
+        while (numBytesPerCheckpointEntry + nextSmrEntrySize < batchThresholdPercentage * getRuntime().getParameters().getMaxWriteSize()) {
+            m.put(keyPrefix + i, payload);
+            payload = getRandomStringOfSize(1 << 20);
+            numBytesPerCheckpointEntry += nextSmrEntrySize;
+            nextSmrEntrySize = getSerializedSMREntrySize(keyPrefix + i, payload,
+                    cpw.getKeyMutator(), cpw.getValueMutator(), cpw.getSerializer());
+            System.out.println(String.format("After putting entry %s, numBytesPerCheckpointEntry is: %s",
+                    (keyPrefix + i), numBytesPerCheckpointEntry));
+            System.out.println("nextSmrEntrySize: " + nextSmrEntrySize);
+            System.out.println("---------------------------------");
```

Comment:
Don;t use logging statements. Replace them with assertions.

Reply:
These prints are for debugging only. I'll just remove them.